src/watcherEx.ts

```diff
@@ -0,0 +1,252 @@
+   * newWatcherEx({
+   *   port: 6379, // Redis port
+   *   host: "127.0.0.1", // Redis host
+   *   family: 4, // 4 (IPv4) or 6 (IPv6)
+   *   password: "auth",
+   *   db: 0,
+   *   channel: "casbin"
+   * })
+   */
+  public static async newWatcherEx(
+    callBackOptions: WatcherExCallBackOptions,
+    options?: WatcherExOptions | string
+  ): Promise<RedisWatcherEx> {
+    let channel = '';
+    if (typeof options === 'object' && options.channel) {
+      channel = options.channel;
+    }
+
+    const pubConnection = new RedisConnection(options);
+    const subConnection = new RedisConnection(options);
+    return this.init(callBackOptions, pubConnection, subConnection, channel);
+  }
+
+  /**
+   * newWatcherExWithCluster creates a watcherEx on the Redis cluster.
+   * @param nodes - An array of nodes in the cluster, [{ port: number, host: string }]
+   * @param clusterOptions - https://github.com/luin/ioredis/blob/master/API.md#new-clusterstartupnodes-options
+   * @example
+   * newWatcherExWithCluster([{ port: 6380, host: "127.0.0.1"}, { port: 6381, host: "127.0.0.1"})
+   */
+  public static async newWatcherExWithCluster(
+    nodes: ClusterNode[] = [],
+    clusterOptions: WatcherExClusterOptions = {},
+    callBackOptions: WatcherExCallBackOptions
+  ): Promise<RedisWatcherEx> {
+    const pubConnection = new RedisClusterConnection(nodes, clusterOptions);
+    const subConnection = new RedisClusterConnection(nodes, clusterOptions);
+    return this.init(callBackOptions, pubConnection, subConnection, clusterOptions.channel);
+  }
+
+  private static async init(
+    callBackOptions: WatcherExCallBackOptions,
+    pubConnection: RedisConnection | RedisClusterConnection,
+    subConnection: RedisConnection | RedisClusterConnection,
+    channel?: string
+  ): Promise<RedisWatcherEx> {
+    const watcherEx = new RedisWatcherEx();
+
+    watcherEx.pubConnection = pubConnection;
+    watcherEx.subConnection = subConnection;
+
+    if (channel) {
+      watcherEx.channel = channel;
+    }
+
+    watcherEx.watcherExCallBackOptions = callBackOptions;
+    const watcherExCallBackMethods: any = {
+      updateForAddPolicy: callBackOptions.addPolicyCb,
+      updateForRemovePolicy: callBackOptions.removePolicyCb,
+      updateForRemoveFilteredPolicy: callBackOptions.removeFilteredPolicyCb,
+      updateForSavePolicy: callBackOptions.savePolicyCb,
+      updateForAddPolicies: callBackOptions.addPoliciesCb,
+      updateForRemovePolicies: callBackOptions.removePoliciesCb
+    };
+
+    const client = await watcherEx.subConnection.getRedisClient();
+    // @ts-ignore - subscribe is exists.
+    await client.subscribe(watcherEx.channel);
+    client.on('message', (chan: string, msg: any) => {
+      if (chan !== watcherEx.channel) {
+        return;
+      }
+      msg = JSON.parse(JSON.stringify(msg));
+      switch (msg.updateType) {
+        case 'UpdateForAddPolicy':
+          watcherExCallBackMethods.updateForAddPolicy(msg.args);
```

Comment:
Currently, the interface allows to not define these functions, but receiving a message from them would throw an error. I think the interface should guide users to not get unnecessary errors. Not defining them should be deliberate, so they should use `@ts-ignore` at that point if they really want to do so.

Reply:
Or there should be an option to not receive errors if a callback is not defined. Either or.